flash/image/classification/adapters.py

```diff
@@ -0,0 +1,518 @@
+        **kwargs,
+    ) -> Adapter:
+        if "meta_batch_size" not in kwargs:
+            raise MisconfigurationException(
+                "The `meta_batch_size` should be provided as training_strategy_kwargs={'meta_batch_size'=...}. "
+                "This is equivalent to the epoch length."
+            )
+        if "shots" not in kwargs:
+            raise MisconfigurationException(
+                "The `shots` should be provided training_strategy_kwargs={'shots'=...}. "
+                "This is equivalent to the number of sample per label to select within a task."
+            )
+        return cls(task, backbone, head, algorithm, **kwargs)
+
+    def training_step(self, batch, batch_idx) -> Any:
+        input = (batch[DefaultDataKeys.INPUT], batch[DefaultDataKeys.TARGET])
+        return self.model.training_step(input, batch_idx)
+
+    def validation_step(self, batch, batch_idx):
+        # Should be True only for trainer.validate
+        if self.trainer.state.fn == TrainerFn.VALIDATING:
+            self._algorithm_has_validated = True
+        input = (batch[DefaultDataKeys.INPUT], batch[DefaultDataKeys.TARGET])
+        return self.model.validation_step(input, batch_idx)
+
+    def validation_epoch_end(self, outpus: Any):
+        self.model.validation_epoch_end(outpus)
+
+    def test_step(self, batch, batch_idx):
+        input = (batch[DefaultDataKeys.INPUT], batch[DefaultDataKeys.TARGET])
+        return self.model.test_step(input, batch_idx)
+
+    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Any:
+        return self.model.predict_step(batch[DefaultDataKeys.INPUT], batch_idx, dataloader_idx=dataloader_idx)
+
+    def _sanetize_batch_size(self, batch_size: int) -> int:
+        if batch_size != 1:
+            warning_cache.warn(
+                "When using a meta-learning training_strategy, the batch_size should be set to 1. "
+                "HINT: You can modify the `meta_batch_size` to 100 for example by doing "
+                f"{type(self.task)}" + "(training_strategies_kwargs={'meta_batch_size': 100})"
+            )
+        return 1
+
+    def process_train_dataset(
+        self,
+        dataset: BaseAutoDataset,
+        trainer: flash.Trainer,
+        batch_size: int,
+        num_workers: int,
+        pin_memory: bool,
+        collate_fn: Callable,
+        shuffle: bool,
+        drop_last: bool,
+        sampler: Optional[Sampler],
+    ) -> DataLoader:
+        dataset = self._convert_dataset(
+            trainer=trainer,
+            dataset=dataset,
+            ways=self.ways,
+            shots=self.shots,
+            queries=self.queries,
+            num_workers=num_workers,
+        )
+        if isinstance(dataset, IterableDataset):
+            shuffle = False
+            sampler = None
+        return super().process_train_dataset(
+            dataset,
+            trainer,
+            self._sanetize_batch_size(batch_size),
```

Comment:
EDIT: fine with this as this should be handled in the backend and at the moment l2l cannot deal with this for protonet. [Here](https://github.com/learnables/learn2learn/blob/6b741028f0812ba73da58acf7206d93654dde97a/learn2learn/algorithms/lightning/lightning_protonet.py#L115-L141) it always calls [`self.classifier.fit_`](https://github.com/learnables/learn2learn/blob/6b741028f0812ba73da58acf7206d93654dde97a/learn2learn/nn/protonet.py#L81-L100) on the support set (i.e., it is not batched). Please feel free to correct me if I'm wrong -- I'd really love having the possibility to set the batch_size within tasks so I really wish I'm wrong ahah

----

This is cleaner, however I think that having the possibility to set the `batch_size` normally in the datamodule is advantageous.

For example, assume that I specify that I want 1000 examples per class (i.e., `shots + queries = 1000`) -- I might want to do meta-learning not just in few-shots learning scenarios --, but I can process 16 examples at a time on my GPU. I would ideally want to be able to set the batch_size normally in the dataloader and automatically have the model process just 16 samples at a time in each task. In other words, ideally the process should look like

```python
for task in tasks:
    for batch in task:
        ...
```

Though, I don't know how easy it would be to implement

Reply:
The batch_size is used for the dataloader, so it should be always 1.

However, the task batch_size is defined as you shared by `shots + queries`.

To enable large values, it might be possible to perform micro batching directly within the `training_step`.

```py
embeddings = []
for chunked_task in self.chunk_task(task):
    embeddings.append(self.features(chunked_task))

...
```

But I believe there might be some better way to do this, I will think about it.

Reply:
Micro-batching is very useful especially in those case in which one wants to use meta-learning to adapt a model on a downstream task without the "few-shots" limitations